### PR TITLE
refactor: add interceptor after Influxdb lines are converted to grpc row insert

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

... so that other projects could modify the row insert according to their specific intention.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **Refactor**
  - Improved the initialization sequence for frontend services, simplifying the method calls and argument passing for better performance and maintainability.

These changes enhance the backend structure without affecting user-facing features directly.

### CODE REVIEW

**Comment by killme2008** [comment]
@MichaelScofield It's better to use an option to control this behavior and it's disabled by default.

**Comment by MichaelScofield** [comment]
@killme2008 The current behavior is unaffected, new influxdb tables are still nanosecond time index. This PR simply brings compatibility back, so I don't think there's a need for an option.

**Comment by killme2008** [comment]
> @killme2008 The current behavior is unaffected, new influxdb tables are still nanosecond time index. This PR simply brings compatibility back, so I don't think there's a need for an option.

Alignment and checking typically impact performance, often unnecessary.

In other words, the user may create a table with a millisecond time index, and ingestion with influxdb, which we want to prohibit.

**Comment by killme2008** [comment]
We already discussed this before and have an agreement.  Maybe you can discuss it directly with @fengjiachun . I have to sleep.

**Comment by MichaelScofield** [comment]
@killme2008 I have added an option for that.

File: src/servers/src/influxdb.rs [hunk 1]
```diff
@@ -86,12 +93,34 @@ impl TryFrom<InfluxdbRequest> for RowInsertRequests {
             });
             row_writer::write_fields(table_data, fields, &mut one_row)?;
 
-            // timestamp
-            row_writer::write_ts_to_nanos(
+            // Align to the table's time index unit if needed.
+            let target_unit = if auto_align_precision {
+                catalog_manager
```
Comment (fengjiachun) [inline src/servers/src/influxdb.rs:98]: I'm unsure if introducing the catalog here is a good idea.

File: src/servers/src/influxdb.rs [hunk 2]
```diff
@@ -132,7 +162,11 @@ monitor2,host=host4 cpu=66.3,memory=1029 1663840496400340003";
             lines: lines.to_string(),
         };
 
-        let requests: RowInsertRequests = influxdb_req.try_into().unwrap();
+        let catalog_manager = MemoryCatalogManager::new() as _;
```
Comment (killme2008) [inline src/servers/src/influxdb.rs:165]: It's better to add a test to cover `auto_align_precision=true`. You can use a mock catalog manager like `MemoryCatalogManager`.

File: tests-integration/src/cluster.rs [hunk 3]
```diff
@@ -376,14 +377,22 @@ impl GreptimeDbClusterBuilder {
             Arc::new(InvalidateTableCacheHandler::new(cache_registry.clone())),
         ]);
 
+        let options = FrontendOptions {
+            influxdb: InfluxdbOptions {
+                enable: true,
```
Comment (killme2008) [inline tests-integration/src/cluster.rs:382]: It's better to cover the old behavior too.

File: src/servers/src/influxdb.rs (hunk 1 above)
Comment (killme2008) [reply src/servers/src/influxdb.rs:98]: Agree. But I don't have a better solution. Maybe adds some handler to breaking the strong coupling, just like:

```rust
pub struct AlignPrecisionHandler {
   caatlog_manager: CatalogManagerRef,
}
impl AlignPrecisionHandler {
   pub async fn get_target_time_unit(query_ctx: &QueryContextRef, table_name: &str) -> Result<TimeUnit> {
        ....
   }
}

....
to_row_insert_requests(AlignPrecisionHandler{catalog_manager}, ...);


```

File: tests-integration/src/tests.rs [hunk 4]
```diff
@@ -15,7 +15,7 @@
 mod instance_kafka_wal_test;
 mod instance_test;
 mod promql_test;
-mod test_util;
+pub mod test_util;
```
Comment (killme2008) [inline tests-integration/src/tests.rs:18]:
```suggestion
pub(crate) mod test_util;
```

**Review by killme2008 — APPROVED** [review]
LGTM. A better way.

**Review by tisonkun — APPROVED** [review]

**Pull request merged**